{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a66f4074-31a9-485f-86d1-8746010ce40f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Please type your DNA sequence: AAAAAGACNNTGAAAAANNNNT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid Sequence: KKK\n"
     ]
    }
   ],
   "source": [
    "codon_chart = {'AUG': 'M', 'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP', \n",
    "               'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
    "               'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
    "              'UAU': 'Y', 'UAC': 'Y', 'UGU': 'C', 'UGC': 'C', 'UGG': 'W',\n",
    "              'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
    "              'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "              'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "              'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "              'AUU': 'I', 'AUC': 'I', 'AUA': 'I',\n",
    "              'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "              'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "              'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "              'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
    "              'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "              'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "              'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'}\n",
    "\n",
    "dna_seq = input('Please type your DNA sequence:')\n",
    "def dnaseq_aminoacidseq(dna_seq):\n",
    "    rna_seq = dna_seq.replace('T', 'U') # Convert to mRNA seq\n",
    "    amino_acids = []\n",
    "    for i in range(0, len(rna_seq), 3): # Range from 1st to last index, in steps of 3 (bc every 3 nucleotides is 1 codon)\n",
    "        codon = rna_seq[i:i+3]\n",
    "        if codon in codon_chart:\n",
    "            amino_acid = codon_chart[codon]\n",
    "            if amino_acid == 'STOP':\n",
    "                break\n",
    "            amino_acids.append(amino_acid)\n",
    "    amino_acids_string = ''.join([str(element) for element in amino_acids]) # Converting to string\n",
    "    return amino_acids_string\n",
    "\n",
    "amino_acid_seq = dnaseq_aminoacidseq(dna_seq)\n",
    "print(f\"Amino Acid Sequence: {amino_acid_seq}\")\n",
    "\n",
    "# Can use .transcribe() and .translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "43053ea0-5529-47cb-b6fd-11cbbac3f472",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query : Query_3532867 Length: 591 Strand: Plus\n",
      "        unnamed protein product\n",
      "Target: pdb|1A35|A Length: 591 Strand: Plus\n",
      "        Chain A, PROTEIN (DNA TOPOISOMERASE I) [Homo sapiens] >pdb|1A36|A Chain\n",
      "        A, TOPOISOMERASE I [Homo sapiens]\n",
      "\n",
      "Score:1216 bits(3145), Expect:0,\n",
      "Identities:590/591(100%),  Positives:591/591(100%),  Gaps:0.591(0%)\n",
      "\n",
      "pdb|1A35|         0 KPKNKDKDKKVPEPDNKKKKPKKEEEQKWKWWEEERYPEGIKWKFLEHKGPVFAPPYEPL\n",
      "                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353         0 KPKNKDKDKKVPEPDNKKKKPKKEEEQKWKWWEEERYPEGIKWKFLEHKGPVFAPPYEPL\n",
      "\n",
      "pdb|1A35|        60 PENVKFYYDGKVMKLSPKAEEVATFFAKMLDHEYTTKEIFRKNFFKDWRKEMTNEEKNII\n",
      "                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353        60 PENVKFYYDGKVMKLSPKAEEVATFFAKMLDHEYTTKEIFRKNFFKDWRKEMTNEEKNII\n",
      "\n",
      "pdb|1A35|       120 TNLSKCDFTQMSQYFKAQTEARKQMSKEEKLKIKEENEKLLKEYGFCIMDNHKERIANFK\n",
      "                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       120 TNLSKCDFTQMSQYFKAQTEARKQMSKEEKLKIKEENEKLLKEYGFCIMDNHKERIANFK\n",
      "\n",
      "pdb|1A35|       180 IEPPGLFRGRGNHPKMGMLKRRIMPEDIIINCSKDAKVPSPPPGHKWKEVRHDNKVTWLV\n",
      "                180 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       180 IEPPGLFRGRGNHPKMGMLKRRIMPEDIIINCSKDAKVPSPPPGHKWKEVRHDNKVTWLV\n",
      "\n",
      "pdb|1A35|       240 SWTENIQGSIKYIMLNPSSRIKGEKDWQKYETARRLKKCVDKIRNQYREDWKSKEMKVRQ\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       240 SWTENIQGSIKYIMLNPSSRIKGEKDWQKYETARRLKKCVDKIRNQYREDWKSKEMKVRQ\n",
      "\n",
      "pdb|1A35|       300 RAVALYFIDKLALRAGNEKEEGETADTVGCCSLRVEHINLHPELDGQEYVVEFDFLGKDS\n",
      "                300 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       300 RAVALYFIDKLALRAGNEKEEGETADTVGCCSLRVEHINLHPELDGQEYVVEFDFLGKDS\n",
      "\n",
      "pdb|1A35|       360 IRYYNKVPVEKRVFKNLQLFMENKQPEDDLFDRLNTGILNKHLQDLMEGLTAKVFRTYNA\n",
      "                360 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       360 IRYYNKVPVEKRVFKNLQLFMENKQPEDDLFDRLNTGILNKHLQDLMEGLTAKVFRTYNA\n",
      "\n",
      "pdb|1A35|       420 SITLQQQLKELTAPDENIPAKILSYNRANRAVAILCNHQRAPPKTFEKSMMNLQTKIDAK\n",
      "                420 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       420 SITLQQQLKELTAPDENIPAKILSYNRANRAVAILCNHQRAPPKTFEKSMMNLQTKIDAK\n",
      "\n",
      "pdb|1A35|       480 KEQLADARRDLKSAKADAKVMKDAKTKKVVESKKKAVQRLEEQLMKLEVQATDREENKQI\n",
      "                480 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "Query_353       480 KEQLADARRDLKSAKADAKVMKDAKTKKVVESKKKAVQRLEEQLMKLEVQATDREENKQI\n",
      "\n",
      "pdb|1A35|       540 ALGTSKLNFLDPRITVAWCKKWGVPIEKIYNKTQREKFAWAIDMADEDYEF 591\n",
      "                540 ||||||||.|||||||||||||||||||||||||||||||||||||||||| 591\n",
      "Query_353       540 ALGTSKLNYLDPRITVAWCKKWGVPIEKIYNKTQREKFAWAIDMADEDYEF 591\n",
      "\n",
      " \n",
      "\n",
      " 1A35_A\n"
     ]
    }
   ],
   "source": [
    "from Bio import Blast # Blast module from Biopython\n",
    "\n",
    "result_stream = Blast.qblast('blastp', 'pdb', 'KPKNKDKDKKVPEPDNKKKKPKKEEEQKWKWWEEERYPEGIKWKFLEHKGPVFAPPYEPLPENVKFYYDGKVMKLSPKAEEVATFFAKMLDHEYTTKEIFRKNFFKDWRKEMTNEEKNIITNLSKCDFTQMSQYFKAQTEARKQMSKEEKLKIKEENEKLLKEYGFCIMDNHKERIANFKIEPPGLFRGRGNHPKMGMLKRRIMPEDIIINCSKDAKVPSPPPGHKWKEVRHDNKVTWLVSWTENIQGSIKYIMLNPSSRIKGEKDWQKYETARRLKKCVDKIRNQYREDWKSKEMKVRQRAVALYFIDKLALRAGNEKEEGETADTVGCCSLRVEHINLHPELDGQEYVVEFDFLGKDSIRYYNKVPVEKRVFKNLQLFMENKQPEDDLFDRLNTGILNKHLQDLMEGLTAKVFRTYNASITLQQQLKELTAPDENIPAKILSYNRANRAVAILCNHQRAPPKTFEKSMMNLQTKIDAKKEQLADARRDLKSAKADAKVMKDAKTKKVVESKKKAVQRLEEQLMKLEVQATDREENKQIALGTSKLNYLDPRITVAWCKKWGVPIEKIYNKTQREKFAWAIDMADEDYEF')\n",
    "# qblast passes sequence data as the main parameter\n",
    "# Internal programme: blastp for protein. Otherwise, blastn is for nucleotides\n",
    "# Database: pdb to access pdb files only\n",
    "# Database: nr stands for non-redundant. All identical sequences are represented by a single entry in the database. Allows to access all identical sequences in the database\n",
    "# Results are given in XML format by default\n",
    "\n",
    "with open(\"my_blast.xml\", \"wb\") as out_stream:\n",
    "    out_stream.write(result_stream.read())\n",
    "# Results are read and written to a file\n",
    "# wb: File is opened in write-binary mode\n",
    "\n",
    "result_stream = open(\"my_blast.xml\", \"rb\")\n",
    "# rb: File is opened in read-binary mode\n",
    "\n",
    "blast_record = Blast.read(result_stream)\n",
    "# File is opened and read in BLAST\n",
    "\n",
    "print(blast_record[0][0], # # Access the first record (though only one query is input in this example) and first hit (alignment)\n",
    "      '\\n\\n', blast_record[0].target.name) # Sequence name/ identifier \n",
    "# There are multiple hits in 1 record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "49d5ccb4-57d7-4d50-9fe7-4c6cca071ec0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1477f9b3934d45499ec6dd6edd05cac0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "pdbid = blast_record[0].target.name.split('_')[0]\n",
    "from biopandas.pdb import PandasPdb\n",
    "ppdb = PandasPdb().fetch_pdb(pdbid) # because id also specifies the chain but pdb cannot read it so just taking the id of the whole protein\n",
    "f'{pdbid}.pdb'\n",
    "file_path = f'{pdbid}.pdb'\n",
    "ppdb.to_pdb(path = file_path)\n",
    "view = nv.show_file(file_path)\n",
    "view.add_representation('cartoon')\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a5e3c769-ca54-4226-8139-c895f64a0756",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 5313.\n",
      "  warnings.warn(\n",
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 5336.\n",
      "  warnings.warn(\n",
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5380.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 1 has secondary structure: -\n",
      "Residue 2 has secondary structure: -\n",
      "Residue 3 has secondary structure: -\n",
      "Residue 4 has secondary structure: S\n",
      "Residue 5 has secondary structure: -\n",
      "Residue 6 has secondary structure: E\n",
      "Residue 7 has secondary structure: E\n",
      "Residue 8 has secondary structure: E\n",
      "Residue 9 has secondary structure: -\n",
      "Residue 10 has secondary structure: -\n",
      "Residue 11 has secondary structure: -\n",
      "Residue 12 has secondary structure: B\n",
      "Residue 13 has secondary structure: -\n",
      "Residue 14 has secondary structure: -\n",
      "Residue 15 has secondary structure: -\n",
      "Residue 16 has secondary structure: -\n",
      "Residue 17 has secondary structure: -\n",
      "Residue 18 has secondary structure: -\n",
      "Residue 19 has secondary structure: -\n",
      "Residue 20 has secondary structure: -\n",
      "Residue 21 has secondary structure: -\n",
      "Residue 22 has secondary structure: S\n",
      "Residue 23 has secondary structure: S\n",
      "Residue 24 has secondary structure: S\n",
      "Residue 25 has secondary structure: -\n",
      "Residue 26 has secondary structure: -\n",
      "Residue 27 has secondary structure: E\n",
      "Residue 28 has secondary structure: E\n",
      "Residue 29 has secondary structure: T\n",
      "Residue 30 has secondary structure: T\n",
      "Residue 31 has secondary structure: E\n",
      "Residue 32 has secondary structure: E\n",
      "Residue 33 has secondary structure: -\n",
      "Residue 34 has secondary structure: -\n",
      "Residue 35 has secondary structure: -\n",
      "Residue 36 has secondary structure: -\n",
      "Residue 37 has secondary structure: H\n",
      "Residue 38 has secondary structure: H\n",
      "Residue 39 has secondary structure: H\n",
      "Residue 40 has secondary structure: H\n",
      "Residue 41 has secondary structure: H\n",
      "Residue 42 has secondary structure: H\n",
      "Residue 43 has secondary structure: H\n",
      "Residue 44 has secondary structure: H\n",
      "Residue 45 has secondary structure: H\n",
      "Residue 46 has secondary structure: H\n",
      "Residue 47 has secondary structure: H\n",
      "Residue 48 has secondary structure: T\n",
      "Residue 49 has secondary structure: T\n",
      "Residue 50 has secondary structure: T\n",
      "Residue 51 has secondary structure: T\n",
      "Residue 52 has secondary structure: S\n",
      "Residue 53 has secondary structure: G\n",
      "Residue 54 has secondary structure: G\n",
      "Residue 55 has secondary structure: G\n",
      "Residue 56 has secondary structure: G\n",
      "Residue 57 has secondary structure: S\n",
      "Residue 58 has secondary structure: H\n",
      "Residue 59 has secondary structure: H\n",
      "Residue 60 has secondary structure: H\n",
      "Residue 61 has secondary structure: H\n",
      "Residue 62 has secondary structure: H\n",
      "Residue 63 has secondary structure: H\n",
      "Residue 64 has secondary structure: H\n",
      "Residue 65 has secondary structure: H\n",
      "Residue 66 has secondary structure: H\n",
      "Residue 67 has secondary structure: H\n",
      "Residue 68 has secondary structure: H\n",
      "Residue 69 has secondary structure: H\n",
      "Residue 70 has secondary structure: T\n",
      "Residue 71 has secondary structure: T\n",
      "Residue 72 has secondary structure: S\n",
      "Residue 73 has secondary structure: -\n",
      "Residue 74 has secondary structure: T\n",
      "Residue 75 has secondary structure: T\n",
      "Residue 76 has secondary structure: H\n",
      "Residue 77 has secondary structure: H\n",
      "Residue 78 has secondary structure: H\n",
      "Residue 79 has secondary structure: H\n",
      "Residue 80 has secondary structure: -\n",
      "Residue 81 has secondary structure: -\n",
      "Residue 82 has secondary structure: -\n",
      "Residue 83 has secondary structure: G\n",
      "Residue 84 has secondary structure: G\n",
      "Residue 85 has secondary structure: G\n",
      "Residue 86 has secondary structure: E\n",
      "Residue 87 has secondary structure: E\n",
      "Residue 88 has secondary structure: -\n",
      "Residue 89 has secondary structure: H\n",
      "Residue 90 has secondary structure: H\n",
      "Residue 91 has secondary structure: H\n",
      "Residue 92 has secondary structure: H\n",
      "Residue 93 has secondary structure: H\n",
      "Residue 94 has secondary structure: H\n",
      "Residue 95 has secondary structure: H\n",
      "Residue 96 has secondary structure: H\n",
      "Residue 97 has secondary structure: H\n",
      "Residue 98 has secondary structure: H\n",
      "Residue 99 has secondary structure: H\n",
      "Residue 100 has secondary structure: H\n",
      "Residue 101 has secondary structure: H\n",
      "Residue 102 has secondary structure: H\n",
      "Residue 103 has secondary structure: H\n",
      "Residue 104 has secondary structure: T\n",
      "Residue 105 has secondary structure: -\n",
      "Residue 106 has secondary structure: -\n",
      "Residue 107 has secondary structure: H\n",
      "Residue 108 has secondary structure: H\n",
      "Residue 109 has secondary structure: H\n",
      "Residue 110 has secondary structure: H\n",
      "Residue 111 has secondary structure: H\n",
      "Residue 112 has secondary structure: H\n",
      "Residue 113 has secondary structure: H\n",
      "Residue 114 has secondary structure: H\n",
      "Residue 115 has secondary structure: H\n",
      "Residue 116 has secondary structure: H\n",
      "Residue 117 has secondary structure: H\n",
      "Residue 118 has secondary structure: H\n",
      "Residue 119 has secondary structure: H\n",
      "Residue 120 has secondary structure: H\n",
      "Residue 121 has secondary structure: H\n",
      "Residue 122 has secondary structure: H\n",
      "Residue 123 has secondary structure: H\n",
      "Residue 124 has secondary structure: H\n",
      "Residue 125 has secondary structure: S\n",
      "Residue 126 has secondary structure: E\n",
      "Residue 127 has secondary structure: E\n",
      "Residue 128 has secondary structure: E\n",
      "Residue 129 has secondary structure: E\n",
      "Residue 130 has secondary structure: T\n",
      "Residue 131 has secondary structure: T\n",
      "Residue 132 has secondary structure: E\n",
      "Residue 133 has secondary structure: E\n",
      "Residue 134 has secondary structure: E\n",
      "Residue 135 has secondary structure: E\n",
      "Residue 136 has secondary structure: B\n",
      "Residue 137 has secondary structure: S\n",
      "Residue 138 has secondary structure: -\n",
      "Residue 139 has secondary structure: -\n",
      "Residue 140 has secondary structure: B\n",
      "Residue 141 has secondary structure: -\n",
      "Residue 142 has secondary structure: -\n",
      "Residue 143 has secondary structure: -\n",
      "Residue 144 has secondary structure: S\n",
      "Residue 145 has secondary structure: E\n",
      "Residue 146 has secondary structure: E\n",
      "Residue 147 has secondary structure: -\n",
      "Residue 148 has secondary structure: -\n",
      "Residue 149 has secondary structure: -\n",
      "Residue 150 has secondary structure: S\n",
      "Residue 151 has secondary structure: T\n",
      "Residue 152 has secondary structure: T\n",
      "Residue 153 has secondary structure: -\n",
      "Residue 154 has secondary structure: T\n",
      "Residue 155 has secondary structure: T\n",
      "Residue 156 has secondary structure: T\n",
      "Residue 157 has secondary structure: T\n",
      "Residue 158 has secondary structure: -\n",
      "Residue 159 has secondary structure: E\n",
      "Residue 160 has secondary structure: E\n",
      "Residue 161 has secondary structure: -\n",
      "Residue 162 has secondary structure: -\n",
      "Residue 163 has secondary structure: -\n",
      "Residue 164 has secondary structure: -\n",
      "Residue 165 has secondary structure: G\n",
      "Residue 166 has secondary structure: G\n",
      "Residue 167 has secondary structure: G\n",
      "Residue 168 has secondary structure: -\n",
      "Residue 169 has secondary structure: E\n",
      "Residue 170 has secondary structure: E\n",
      "Residue 171 has secondary structure: E\n",
      "Residue 172 has secondary structure: -\n",
      "Residue 173 has secondary structure: -\n",
      "Residue 174 has secondary structure: T\n",
      "Residue 175 has secondary structure: T\n",
      "Residue 176 has secondary structure: S\n",
      "Residue 177 has secondary structure: -\n",
      "Residue 178 has secondary structure: -\n",
      "Residue 179 has secondary structure: -\n",
      "Residue 180 has secondary structure: -\n",
      "Residue 181 has secondary structure: -\n",
      "Residue 182 has secondary structure: S\n",
      "Residue 183 has secondary structure: T\n",
      "Residue 184 has secondary structure: T\n",
      "Residue 185 has secondary structure: -\n",
      "Residue 186 has secondary structure: -\n",
      "Residue 187 has secondary structure: -\n",
      "Residue 188 has secondary structure: S\n",
      "Residue 189 has secondary structure: E\n",
      "Residue 190 has secondary structure: E\n",
      "Residue 191 has secondary structure: E\n",
      "Residue 192 has secondary structure: -\n",
      "Residue 193 has secondary structure: -\n",
      "Residue 194 has secondary structure: T\n",
      "Residue 195 has secondary structure: T\n",
      "Residue 196 has secondary structure: -\n",
      "Residue 197 has secondary structure: -\n",
      "Residue 198 has secondary structure: -\n",
      "Residue 199 has secondary structure: S\n",
      "Residue 200 has secondary structure: E\n",
      "Residue 201 has secondary structure: E\n",
      "Residue 202 has secondary structure: E\n",
      "Residue 203 has secondary structure: E\n",
      "Residue 204 has secondary structure: -\n",
      "Residue 205 has secondary structure: T\n",
      "Residue 206 has secondary structure: T\n",
      "Residue 207 has secondary structure: T\n",
      "Residue 208 has secondary structure: -\n",
      "Residue 209 has secondary structure: -\n",
      "Residue 210 has secondary structure: E\n",
      "Residue 211 has secondary structure: E\n",
      "Residue 212 has secondary structure: E\n",
      "Residue 213 has secondary structure: E\n",
      "Residue 214 has secondary structure: -\n",
      "Residue 215 has secondary structure: B\n",
      "Residue 216 has secondary structure: -\n",
      "Residue 217 has secondary structure: T\n",
      "Residue 218 has secondary structure: T\n",
      "Residue 219 has secondary structure: S\n",
      "Residue 220 has secondary structure: H\n",
      "Residue 221 has secondary structure: H\n",
      "Residue 222 has secondary structure: H\n",
      "Residue 223 has secondary structure: H\n",
      "Residue 224 has secondary structure: H\n",
      "Residue 225 has secondary structure: H\n",
      "Residue 226 has secondary structure: H\n",
      "Residue 227 has secondary structure: H\n",
      "Residue 228 has secondary structure: H\n",
      "Residue 229 has secondary structure: H\n",
      "Residue 230 has secondary structure: H\n",
      "Residue 231 has secondary structure: H\n",
      "Residue 232 has secondary structure: H\n",
      "Residue 233 has secondary structure: H\n",
      "Residue 234 has secondary structure: H\n",
      "Residue 235 has secondary structure: H\n",
      "Residue 236 has secondary structure: H\n",
      "Residue 237 has secondary structure: H\n",
      "Residue 238 has secondary structure: H\n",
      "Residue 239 has secondary structure: H\n",
      "Residue 240 has secondary structure: H\n",
      "Residue 241 has secondary structure: H\n",
      "Residue 242 has secondary structure: H\n",
      "Residue 243 has secondary structure: H\n",
      "Residue 244 has secondary structure: H\n",
      "Residue 245 has secondary structure: H\n",
      "Residue 246 has secondary structure: H\n",
      "Residue 247 has secondary structure: H\n",
      "Residue 248 has secondary structure: H\n",
      "Residue 249 has secondary structure: H\n",
      "Residue 250 has secondary structure: H\n",
      "Residue 251 has secondary structure: T\n",
      "Residue 252 has secondary structure: T\n",
      "Residue 253 has secondary structure: -\n",
      "Residue 254 has secondary structure: S\n",
      "Residue 255 has secondary structure: -\n",
      "Residue 256 has secondary structure: H\n",
      "Residue 257 has secondary structure: H\n",
      "Residue 258 has secondary structure: H\n",
      "Residue 259 has secondary structure: H\n",
      "Residue 260 has secondary structure: H\n",
      "Residue 261 has secondary structure: H\n",
      "Residue 262 has secondary structure: H\n",
      "Residue 263 has secondary structure: H\n",
      "Residue 264 has secondary structure: H\n",
      "Residue 265 has secondary structure: H\n",
      "Residue 266 has secondary structure: H\n",
      "Residue 267 has secondary structure: H\n",
      "Residue 268 has secondary structure: H\n",
      "Residue 269 has secondary structure: H\n",
      "Residue 270 has secondary structure: H\n",
      "Residue 271 has secondary structure: H\n",
      "Residue 272 has secondary structure: -\n",
      "Residue 273 has secondary structure: -\n",
      "Residue 274 has secondary structure: -\n",
      "Residue 275 has secondary structure: -\n",
      "Residue 276 has secondary structure: -\n",
      "Residue 277 has secondary structure: -\n",
      "Residue 278 has secondary structure: -\n",
      "Residue 279 has secondary structure: -\n",
      "Residue 280 has secondary structure: -\n",
      "Residue 281 has secondary structure: T\n",
      "Residue 282 has secondary structure: T\n",
      "Residue 283 has secondary structure: T\n",
      "Residue 284 has secondary structure: S\n",
      "Residue 285 has secondary structure: -\n",
      "Residue 286 has secondary structure: -\n",
      "Residue 287 has secondary structure: -\n",
      "Residue 288 has secondary structure: -\n",
      "Residue 289 has secondary structure: -\n",
      "Residue 290 has secondary structure: S\n",
      "Residue 291 has secondary structure: T\n",
      "Residue 292 has secondary structure: T\n",
      "Residue 293 has secondary structure: -\n",
      "Residue 294 has secondary structure: B\n",
      "Residue 295 has secondary structure: G\n",
      "Residue 296 has secondary structure: G\n",
      "Residue 297 has secondary structure: G\n",
      "Residue 298 has secondary structure: E\n",
      "Residue 299 has secondary structure: E\n",
      "Residue 300 has secondary structure: E\n",
      "Residue 301 has secondary structure: E\n",
      "Residue 302 has secondary structure: E\n",
      "Residue 303 has secondary structure: E\n",
      "Residue 304 has secondary structure: E\n",
      "Residue 305 has secondary structure: T\n",
      "Residue 306 has secondary structure: T\n",
      "Residue 307 has secondary structure: E\n",
      "Residue 308 has secondary structure: E\n",
      "Residue 309 has secondary structure: E\n",
      "Residue 310 has secondary structure: E\n",
      "Residue 311 has secondary structure: E\n",
      "Residue 312 has secondary structure: E\n",
      "Residue 313 has secondary structure: E\n",
      "Residue 314 has secondary structure: E\n",
      "Residue 315 has secondary structure: E\n",
      "Residue 316 has secondary structure: E\n",
      "Residue 317 has secondary structure: -\n",
      "Residue 318 has secondary structure: G\n",
      "Residue 319 has secondary structure: G\n",
      "Residue 320 has secondary structure: G\n",
      "Residue 321 has secondary structure: -\n",
      "Residue 322 has secondary structure: E\n",
      "Residue 323 has secondary structure: E\n",
      "Residue 324 has secondary structure: E\n",
      "Residue 325 has secondary structure: E\n",
      "Residue 326 has secondary structure: E\n",
      "Residue 327 has secondary structure: E\n",
      "Residue 328 has secondary structure: E\n",
      "Residue 329 has secondary structure: -\n",
      "Residue 330 has secondary structure: -\n",
      "Residue 331 has secondary structure: H\n",
      "Residue 332 has secondary structure: H\n",
      "Residue 333 has secondary structure: H\n",
      "Residue 334 has secondary structure: H\n",
      "Residue 335 has secondary structure: H\n",
      "Residue 336 has secondary structure: H\n",
      "Residue 337 has secondary structure: H\n",
      "Residue 338 has secondary structure: H\n",
      "Residue 339 has secondary structure: H\n",
      "Residue 340 has secondary structure: H\n",
      "Residue 341 has secondary structure: H\n",
      "Residue 342 has secondary structure: T\n",
      "Residue 343 has secondary structure: T\n",
      "Residue 344 has secondary structure: -\n",
      "Residue 345 has secondary structure: -\n",
      "Residue 346 has secondary structure: T\n",
      "Residue 347 has secondary structure: T\n",
      "Residue 348 has secondary structure: S\n",
      "Residue 349 has secondary structure: B\n",
      "Residue 350 has secondary structure: T\n",
      "Residue 351 has secondary structure: T\n",
      "Residue 352 has secondary structure: T\n",
      "Residue 353 has secondary structure: T\n",
      "Residue 354 has secondary structure: -\n",
      "Residue 355 has secondary structure: -\n",
      "Residue 356 has secondary structure: H\n",
      "Residue 357 has secondary structure: H\n",
      "Residue 358 has secondary structure: H\n",
      "Residue 359 has secondary structure: H\n",
      "Residue 360 has secondary structure: H\n",
      "Residue 361 has secondary structure: H\n",
      "Residue 362 has secondary structure: H\n",
      "Residue 363 has secondary structure: H\n",
      "Residue 364 has secondary structure: H\n",
      "Residue 365 has secondary structure: H\n",
      "Residue 366 has secondary structure: H\n",
      "Residue 367 has secondary structure: S\n",
      "Residue 368 has secondary structure: T\n",
      "Residue 369 has secondary structure: T\n",
      "Residue 370 has secondary structure: -\n",
      "Residue 371 has secondary structure: -\n",
      "Residue 372 has secondary structure: H\n",
      "Residue 373 has secondary structure: H\n",
      "Residue 374 has secondary structure: H\n",
      "Residue 375 has secondary structure: H\n",
      "Residue 376 has secondary structure: H\n",
      "Residue 377 has secondary structure: H\n",
      "Residue 378 has secondary structure: H\n",
      "Residue 379 has secondary structure: H\n",
      "Residue 380 has secondary structure: H\n",
      "Residue 381 has secondary structure: H\n",
      "Residue 382 has secondary structure: H\n",
      "Residue 383 has secondary structure: H\n",
      "Residue 384 has secondary structure: H\n",
      "Residue 385 has secondary structure: H\n",
      "Residue 386 has secondary structure: H\n",
      "Residue 387 has secondary structure: H\n",
      "Residue 388 has secondary structure: H\n",
      "Residue 389 has secondary structure: H\n",
      "Residue 390 has secondary structure: H\n",
      "Residue 391 has secondary structure: S\n",
      "Residue 392 has secondary structure: -\n",
      "Residue 393 has secondary structure: -\n",
      "Residue 394 has secondary structure: T\n",
      "Residue 395 has secondary structure: T\n",
      "Residue 396 has secondary structure: -\n",
      "Residue 397 has secondary structure: -\n",
      "Residue 398 has secondary structure: H\n",
      "Residue 399 has secondary structure: H\n",
      "Residue 400 has secondary structure: H\n",
      "Residue 401 has secondary structure: H\n",
      "Residue 402 has secondary structure: H\n",
      "Residue 403 has secondary structure: H\n",
      "Residue 404 has secondary structure: H\n",
      "Residue 405 has secondary structure: H\n",
      "Residue 406 has secondary structure: H\n",
      "Residue 407 has secondary structure: H\n",
      "Residue 408 has secondary structure: H\n",
      "Residue 409 has secondary structure: H\n",
      "Residue 410 has secondary structure: H\n",
      "Residue 411 has secondary structure: H\n",
      "Residue 412 has secondary structure: H\n",
      "Residue 413 has secondary structure: H\n",
      "Residue 414 has secondary structure: H\n",
      "Residue 415 has secondary structure: H\n",
      "Residue 416 has secondary structure: T\n",
      "Residue 417 has secondary structure: T\n",
      "Residue 418 has secondary structure: -\n",
      "Residue 419 has secondary structure: B\n",
      "Residue 420 has secondary structure: -\n",
      "Residue 421 has secondary structure: -\n",
      "Residue 423 has secondary structure: -\n",
      "Residue 424 has secondary structure: B\n",
      "Residue 425 has secondary structure: -\n",
      "Residue 426 has secondary structure: -\n",
      "Residue 427 has secondary structure: H\n",
      "Residue 428 has secondary structure: H\n",
      "Residue 429 has secondary structure: I\n",
      "Residue 430 has secondary structure: I\n",
      "Residue 431 has secondary structure: I\n",
      "Residue 432 has secondary structure: I\n",
      "Residue 433 has secondary structure: I\n",
      "Residue 434 has secondary structure: S\n",
      "Residue 435 has secondary structure: -\n",
      "Residue 436 has secondary structure: H\n",
      "Residue 437 has secondary structure: H\n",
      "Residue 438 has secondary structure: H\n",
      "Residue 439 has secondary structure: H\n",
      "Residue 440 has secondary structure: H\n",
      "Residue 441 has secondary structure: H\n",
      "Residue 442 has secondary structure: H\n",
      "Residue 443 has secondary structure: H\n",
      "Residue 444 has secondary structure: H\n",
      "Residue 445 has secondary structure: H\n",
      "Residue 446 has secondary structure: H\n",
      "Residue 447 has secondary structure: T\n",
      "Residue 448 has secondary structure: -\n",
      "Residue 449 has secondary structure: -\n",
      "Residue 450 has secondary structure: H\n",
      "Residue 451 has secondary structure: H\n",
      "Residue 452 has secondary structure: H\n",
      "Residue 453 has secondary structure: H\n",
      "Residue 454 has secondary structure: S\n",
      "Residue 455 has secondary structure: -\n",
      "Residue 456 has secondary structure: H\n",
      "Residue 457 has secondary structure: H\n",
      "Residue 458 has secondary structure: H\n",
      "Residue 459 has secondary structure: H\n",
      "Residue 460 has secondary structure: H\n",
      "Residue 461 has secondary structure: H\n",
      "Residue 462 has secondary structure: G\n",
      "Residue 463 has secondary structure: G\n",
      "Residue 464 has secondary structure: G\n",
      "Residue 465 has secondary structure: H\n",
      "Residue 466 has secondary structure: H\n",
      "Residue 467 has secondary structure: H\n",
      "Residue 468 has secondary structure: H\n",
      "Residue 469 has secondary structure: -\n",
      "Residue 470 has secondary structure: -\n",
      "Residue 471 has secondary structure: T\n",
      "Residue 472 has secondary structure: T\n",
      "Residue 473 has secondary structure: -\n",
      "Residue 474 has secondary structure: -\n",
      "Residue 475 has secondary structure: -\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import PDBParser, DSSP\n",
    "\n",
    "# Load the PDB file\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(pdbid, file_path)\n",
    "\n",
    "# Initialize DSSP\n",
    "dssp = DSSP(structure[0], file_path)\n",
    "\n",
    "# Iterate through the DSSP data to get secondary structure info\n",
    "for key in dssp.keys():\n",
    "    residue = dssp[key]\n",
    "    residue_id = residue[0]\n",
    "    secondary_structure = residue[2]  # Secondary structure code\n",
    "    print(f\"Residue {residue_id} has secondary structure: {secondary_structure}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "370367b0-d221-47a4-80dd-abccf45583dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 5313.\n",
      "  warnings.warn(\n",
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 5336.\n",
      "  warnings.warn(\n",
      "/Users/gabrielle/miniconda3/lib/python3.12/site-packages/Bio/PDB/StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 5380.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fe8aba5e54c44593a8d798323b72f71d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from biopandas.pdb import PandasPdb\n",
    "from Bio.PDB import PDBParser, DSSP\n",
    "\n",
    "# Assume `blast_record` is already defined and contains your PDB ID\n",
    "pdbid = blast_record[0].target.name.split('_')[0]\n",
    "\n",
    "# Fetch the PDB file\n",
    "ppdb = PandasPdb().fetch_pdb(pdbid)\n",
    "file_path = f'{pdbid}.pdb'\n",
    "ppdb.to_pdb(path=file_path)\n",
    "\n",
    "# Load the PDB file with Biopython\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(pdbid, file_path)\n",
    "\n",
    "# Initialize DSSP to get secondary structure assignments\n",
    "dssp = DSSP(structure[0], file_path)\n",
    "\n",
    "# Create a color mapping for secondary structures\n",
    "color_map = {\n",
    "    'H': 'red',    # Alpha helix\n",
    "    'E': 'blue',   # Beta sheet\n",
    "    'C': 'gray',\n",
    "    'B': 'orange',\n",
    "    'S': 'purple',\n",
    "    'G': 'yellow',\n",
    "    'I': 'pink',\n",
    "    'T': 'green'\n",
    "}\n",
    "\n",
    "# Create a list to hold colors for each residue\n",
    "colors = []\n",
    "for key in dssp.keys():\n",
    "    secondary_structure = dssp[key][2]  # Get secondary structure code\n",
    "    color = color_map.get(secondary_structure, 'white')  # Default to white\n",
    "    colors.append(color)\n",
    "\n",
    "# Create an NGLView widget\n",
    "view = nv.show_file(file_path)\n",
    "view.add_representation('cartoon')\n",
    "\n",
    "# Add custom color for each residue based on DSSP assignments\n",
    "for i, color in enumerate(colors):\n",
    "    view.add_representation('stick', selection=f':{i+1}', color=color)\n",
    "\n",
    "# Display the view\n",
    "view"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
